### PR TITLE
Network: change node authentication behavior

### LINKED ISSUE #1859: Alter behaviour of unauthenticated connections
- Mark them as unauthenticated
- Add the state to the diagnostics
- no longer disconnect on unauthenticated

Rereview self-authentication at startup after this issue. (Checking configured NodeDID)

### PR DESCRIPTION
**Pull request by gerardsn**
closes #1859 

Changes in this PR
- removed `DisableNodeAuthentication` config option. (need to check if other repos require changes)
- if authentication fails, the connection is accepted but marked as unauthenticated (upgrading to an authenticated connection will be part of dynamic service discovery)
- documentation
- self-authentication status is now in `/health`. Node startup no longer fails for invalid nodeDID configurations.

`/status/diagnostics` no longer shows a nodeDID when authentication has failed for bootstrap nodes:

```
connected_peers:
    - id: 1e3c49ec-3031-4fdf-8797-91b17aee9776
      address: nuts.nl:5555
      nodedid: ""
      authenticated: false
```

```
connected_peers:
    - id: 1e3c49ec-3031-4fdf-8797-91b17aee9776
      address: nuts.nl:5555
      nodedid: did:nuts:HSr29Y6qFFSHGPzn9VK7dhpbT14U9epwPhMECRHBwda
      authenticated: true
```

### CODE REVIEW

File: network/transport/grpc/connection_manager.go [hunk 1]
```diff
@@ -387,21 +381,21 @@ func (s *grpcConnectionManager) openOutboundStream(connection Connection, protoc
...
+		peer, err = s.authenticator.Authenticate(nodeDID, *peerFromCtx, peer)
 		if err != nil {
 			log.Logger().
 				WithError(err).
 				WithFields(peer.ToFields()).
 				WithField(core.LogFieldDID, nodeDID).
 				Warn("Peer node DID could not be authenticated")
 			// Error message is spec'd by RFC017, because it is returned to the peer
-			return transport.Peer{}, ErrNodeDIDAuthFailed
+			//return transport.Peer{}, ErrNodeDIDAuthFailed // TODO: removing this requires a spec change
```
Comment (reinkrul) [inline network/transport/grpc/connection_manager.go:395]: do we alter the spec in this change? Since this effectively removes it.

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: network/transport/grpc/connection_manager.go (hunk 1 above)
Comment (gerardsn) [reply network/transport/grpc/connection_manager.go:395]: Yes, this change should be backwards compatible though. https://github.com/nuts-foundation/nuts-specification/issues/252

File: network/network.go [hunk 2]
```diff
@@ -307,11 +322,8 @@ func (n *Network) Start() error {
 		return err
 	}
 	if !nodeDID.Empty() {
-		err := n.validateNodeDID(nodeDID)
+		err = n.validateNodeDID(nodeDID)
 		if err != nil {
-			if n.strictMode {
```
[the remark this answers is not in the document]
Comment (gerardsn) [reply network/network.go:312]: I thought this was added after 5.0 and removed before 5.1 (so never in a release). But I see this is only true for the node DID authentication part. 

5.0 already failed when the node does not contain key material for the configured DID, or the DID does not have a NutsComm address.

I would say missing key material should still be fatal, and all other errors are part of authentication so can be in `/health`?

File: network/network.go [hunk 3]
```diff
@@ -221,20 +246,10 @@ func (n *Network) Configure(config core.ServerConfig) error {
...
-		if n.config.DisableNodeAuthentication {
-			// Not allowed in strict mode for security reasons: only intended for demo/workshop purposes.
-			if config.Strictmode {
-				return errors.New("disabling node DID in strict mode is not allowed")
-			}
-			authenticator = grpc.NewDummyAuthenticator(nil)
-		} else {
-			authenticator = grpc.NewTLSAuthenticator(didservice.NewServiceResolver(n.didDocumentResolver))
-		}
+		authenticator := grpc.NewTLSAuthenticator(didservice.NewServiceResolver(n.didDocumentResolver))
```
[the remark this answers is not in the document]
Comment (gerardsn) [reply network/network.go:252]: I did not consider the impact on non-TLS mode. The proposed change still accepts non-TLS connections, but now marks them as unauthenticated. This prevents private transactions exchange in non-TLS mode. Should we allow that anyway?

File: network/network.go [hunk 4]
```diff
@@ -100,6 +100,31 @@ func (n *Network) CheckHealth() map[string]core.Health {
 			}
 		}
 	}
+	// self-authentication
```
[the remark this answers is not in the document]
Comment (gerardsn) [reply network/network.go:103]: Couldn't come up with a good name, but `auth_config` is already better than `auth`. Will change it to that

File: network/network.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply network/network.go:255]: We used it in workshops. So workshop documentation should be checked.

File: docs/pages/deployment/configuration.rst [hunk 5]
```diff
@@ -73,9 +73,8 @@ Save storage of any private key material requires some serious consideration.
 For this reason the ``crypto.storage`` backend must explicitly be set.
 
 Private transactions can only be exchanged over authenticated nodes.
-Therefore strict mode requires that ``network.enabletls=true``,  ``network.disablenodeauthentication=false``, and the certificate chain``tls.{certfile,certkeyfile,truststore}`` must be provided.
-Since node authentication is bi-directional, the node will try to authenticate itself during startup when a ``network.nodedid`` is provided.
-As a consequence, a valid ``NutsComm`` service must be registered to this ``did`` and the address must match that of the provided TLS certificate or the node will not start.
+Therefore strict mode requires that ``network.enabletls=true``, and the certificate chain ``tls.{certfile,certkeyfile,truststore}`` must be provided.
```
Comment (woutslakhorst) [inline docs/pages/deployment/configuration.rst:76]:
```suggestion
Therefore strict mode requires ``network.enabletls=true``, and the certificate chain ``tls.{certfile,certkeyfile,truststore}`` must be provided.
```

File: network/network.go [hunk 6]
```diff
@@ -90,12 +93,37 @@ func (n *Network) CheckHealth() map[string]core.Health {
...
+		}
+	}
+	// auth_config checks that the node is correctly configured to be authenticated by others
+	nodeDID, err := n.nodeDIDResolver.Resolve()
+	if err != nil {
+		// can only happen when not in strictmode and autoNodeDIDResolver fails
+		results[healthAuthConfig] = core.Health{
+			Status:  core.HealthStatusUnknown,
+			Details: err.Error(),
+		}
```
Comment (woutslakhorst) [inline network/network.go:113]: return results here and remove the else wrapper afterwards. Same can be done for the `else if` in the lines below.

File: network/network.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply network/network.go:312]: Could it be possible to mark network connections as partly healthy if one or more outgoing connections are unauthenticated?

**Review by woutslakhorst — CHANGES_REQUESTED** [review]
- add documentation to health/diagnostics in monitoring.rst

File: network/network.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gerardsn) [reply network/network.go:312]: I don't understand what you hope to achieve with this?

The (health)check is to see if the node is configured in such a way that peers can authenticate connections with this node. We do not actually know if the peer has accepted the connection as authenticated. This is the reason to "self-authenticate", so we know that a peer should be able to authenticate this node.

Connections that this node marks as (un)authenticated are the result of the peer's configuration, and therefor not part of the health of this node.

[the remark this answers is not in the document]
Comment (woutslakhorst) [reply network/network.go:312]: Ah yes, you don't know what the other side thinks of you.

File: network/network.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gerardsn) [reply network/network.go:255]: reverted this change. The dummy is used again for non TLS configs

File: echo/echo_integration_test.go [hunk 7]
```diff
@@ -44,7 +44,6 @@ func TestStatusCodes(t *testing.T) {
 	const configFile = `verbosity: debug
 strictmode: false
 network:
-  disablenodeauthentication: true
```
Comment (reinkrul) [inline echo/echo_integration_test.go:47]: I thought this change was reverted?

Comment (gerardsn) [reply echo/echo_integration_test.go:47]: I reinstated the dummy authenticator for disableTLS. Removal of disablenodeauthentication is still the goal of this PR.

**Review by woutslakhorst — APPROVED** [review]

**Review by reinkrul — APPROVED** [review]

**Pull request merged**